stan/math/prim/functor/ode_rk45.hpp

```diff
@@ -0,0 +1,186 @@
+ *   template<typename T_t, typename T_y, typename... T_Args>
+ *   Eigen::Matrix<stan::return_type_t<T_t, T_y, T_Args...>, Eigen::Dynamic, 1>
+ *     operator()(const T_t& t, const Eigen::Matrix<T_y, Eigen::Dynamic, 1>& y,
+ *     std::ostream* msgs, const T_Args&... args);
+ *
+ * t is the time, y is the state, msgs is a stream for error messages, and args
+ * are optional arguments passed to the ODE solve function (which are passed
+ * through to \p f without modification).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_0 Type of initial time
+ * @tparam T_ts Type of output times
+ * @tparam T_Args Types of pass-through parameters
+ *
+ * @param f Right hand side of the ODE
+ * @param y0_arg Initial state
+ * @param t0 Initial time
+ * @param ts Times at which to solve the ODE at. All values must be sorted and
+ *   greater than t0.
+ * @param relative_tolerance Relative tolerance passed to CVODES
+ * @param absolute_tolerance Absolute tolerance passed to CVODES
+ * @param max_num_steps Upper limit on the number of integration steps to
+ *   take between each output (error if exceeded)
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args Extra arguments passed unmodified through to ODE right hand side
+ * @return Solution to ODE at times \p ts
+ */
+template <typename F, typename T_initial, typename T_t0, typename T_ts,
+          typename... Args>
+std::vector<Eigen::Matrix<stan::return_type_t<T_initial, T_t0, T_ts, Args...>,
+                          Eigen::Dynamic, 1>>
+ode_rk45_tol(const F& f,
+             const Eigen::Matrix<T_initial, Eigen::Dynamic, 1>& y0_arg, T_t0 t0,
+             const std::vector<T_ts>& ts, double relative_tolerance,
+             double absolute_tolerance,
+             long int max_num_steps,  // NOLINT(runtime/int)
+             std::ostream* msgs, const Args&... args) {
+  using boost::numeric::odeint::integrate_times;
+  using boost::numeric::odeint::make_dense_output;
+  using boost::numeric::odeint::max_step_checker;
+  using boost::numeric::odeint::runge_kutta_dopri5;
+  using boost::numeric::odeint::vector_space_algebra;
+
+  using T_initial_or_t0 = return_type_t<T_initial, T_t0>;
+
+  Eigen::Matrix<T_initial_or_t0, Eigen::Dynamic, 1> y0 = y0_arg.unaryExpr(
+      [](const T_initial& val) { return T_initial_or_t0(val); });
+
+  check_finite("integrate_ode_rk45", "initial state", y0);
+  check_finite("integrate_ode_rk45", "initial time", t0);
+  check_finite("integrate_ode_rk45", "times", ts);
+
+  // Code from https://stackoverflow.com/a/17340003
+  std::vector<int> unused_temp{
+      0, (check_finite("integrate_ode_rk45", "ode parameters and data", args),
+          0)...};
+
+  check_nonzero_size("integrate_ode_rk45", "initial state", y0);
+  check_nonzero_size("integrate_ode_rk45", "times", ts);
+  check_sorted("integrate_ode_rk45", "times", ts);
+  check_less("integrate_ode_rk45", "initial time", t0, ts[0]);
```

Comment:
I backed off on allowing `t0 == ts[0]` because of the interpretation of the gradients.

`dy/dt0` != `dy/dts[0]`

The one on the left is the derivative with respect to the initial conditions of the ODE problem, and the one on the right is the derivative of the solution of the ODE with fixed initial conditions at different time points.

Now that I type it out I'm not convinced they're different, but I'll double check this.

Reply:
So the thing that I think is confusing is when `t0 == ts[0]`.

I think it's confusing because:

```dy(t = ts[0])/dt0 != dy(t = ts[0])/dts[0]```

and that was not obvious enough that when I read your initial comment the difference didn't immediately pop out to me (not that I'm a super expert, but we aren't targeting exclusively super ODE expert folks).

This is true just from regular 1d integrals -- it's the difference between differentiating between the left hand and right hand limits of the integral (https://en.wikipedia.org/wiki/Leibniz_integral_rule).

I'd like to avoid this confusion by not allowing `ts[0]` to be equal to `t0`.

What do you think?